File: packages/openmemory-js/src/memory/reflect.ts
import { q, log_maint_op, all_async } from "../core/db";
import { add_hsg_memory } from "./hsg";
import { env } from "../core/config";
import { j } from "../utils";

const sim = (t1: string, t2: string): number => {
    const s1 = new Set(
        t1
            .toLowerCase()
            .split(/\s+/)
            .filter((x) => x.length > 0),
    );
    const s2 = new Set(
        t2
            .toLowerCase()
            .split(/\s+/)
            .filter((x) => x.length > 0),
    );
    if (s1.size === 0 || s2.size === 0) return 0;

    let inter = 0;
    for (const token of s1) {
        if (s2.has(token)) inter++;
    }
    const union = new Set([...s1, ...s2]).size;
    return union > 0 ? inter / union : 0;
};

const cluster = (mems: any[]): any[] => {
    const cls: any[] = [];
    const used = new Set();
    for (const m of mems) {
        if (
            used.has(m.id) ||
            m.primary_sector === "reflective" ||
            m.metadata?.consolidated
        )
            continue;
        const c = { mem: [m], n: 1 };
        used.add(m.id);
        for (const o of mems) {
            if (used.has(o.id) || m.primary_sector !== o.primary_sector)
                continue;
            if (sim(m.content, o.content) > 0.8) {
                c.mem.push(o);
                c.n++;
                used.add(o.id);
            }
        }
        if (c.n >= 2) cls.push(c);
    }
    return cls;
};

const sal = (c: any): number => {
    const now = Date.now();
    const p = c.n / 10;
    const r =
        c.mem.reduce(
            (s: number, m: any) =>
                s +
                Math.exp(-(now - new Date(m.created_at).getTime()) / 43200000),
            0,
        ) / c.n;
    const e = c.mem.some(
        (m: any) =>
            m.sectors &&
            Array.isArray(m.sectors) &&
            m.sectors.includes("emotional"),
    )
        ? 1
        : 0;
    return Math.min(1, 0.6 * p + 0.3 * r + 0.1 * e);
};

const summ = (c: any): string => {
    const sec = c.mem[0].primary_sector;
    const n = c.n;
    const txt = c.mem.map((m: any) => m.content.substring(0, 60)).join("; ");
    return `${n} ${sec} pattern: ${txt.substring(0, 200)}`;
};

const mark = async (ids: string[], user_id: string) => {
    const active_user = user_id?.trim();
    if (!active_user) return;
    for (const id of ids) {
        const m = await q.get_mem.get(id);
        if (m && m.user_id === active_user) {
            const meta = JSON.parse(m.meta || "{}");
            meta.consolidated = true;
            await q.upd_mem.run(
                m.content,
                m.tags,
                JSON.stringify(meta),
                Date.now(),
                id,
                active_user,
            );
        }
    }
};

const boost = async (ids: string[], user_id: string) => {
    const active_user = user_id?.trim();
    if (!active_user) return;
    for (const id of ids) {
        const m = await q.get_mem.get(id);
        if (m && m.user_id === active_user) {
            await q.upd_mem.run(m.content, m.tags, m.meta, Date.now(), id, active_user);
            await q.upd_seen.run(
                m.last_seen_at,
                Math.min(1, m.salience * 1.1),
                Date.now(),
                id,
                active_user,
            );
        }
    }
};

export const run_reflection = async (user_id: string, min_override?: number) => {
    const active_user = user_id?.trim();
    if (!active_user) {
        throw new Error("tenant_required: run_reflection requires an authenticated non-empty user_id");
    }

    console.error(`[REFLECT] Starting reflection job...`);
    const min = min_override ?? env.reflect_min ?? 20;
    const mems = await q.all_mem_by_user.all(active_user, 100, 0);
    console.error(
        `[REFLECT] Fetched ${mems.length} memories (min required: ${min})`,
    );
    if (mems.length < min) {
        console.error("[REFLECT] Not enough memories, skipping");
        return { created: 0, reason: "low" };
    }
    const cls = cluster(mems);
    console.error(`[REFLECT] Clustered into ${cls.length} groups`);
    let n = 0;
    for (const c of cls) {
        const txt = summ(c);
        const s = sal(c);
        const src = c.mem.map((m: any) => m.id);
        const meta = {
            type: "auto_reflect",
            sources: src,
            freq: c.n,
            at: new Date().toISOString(),
        };
        console.error(
            `[REFLECT] Creating reflection: ${c.n} memories, salience=${s.toFixed(3)}, sector=${c.mem[0].primary_sector}`,
        );
        await add_hsg_memory(txt, j(["reflect:auto"]), meta, active_user);
        await mark(src, active_user);
        await boost(src, active_user);
        n++;
    }
    if (n > 0) await log_maint_op("reflect", n);
    console.error(`[REFLECT] Job complete: created ${n} reflections`);
    return { created: n, clusters: cls.length };
};

let timer: NodeJS.Timeout | null = null;

export const run_reflection_all_tenants = async (min_override?: number) => {
    const tenant_rows = await all_async(
        "select distinct user_id from memories where user_id is not null and user_id != ''",
    );
    let total_created = 0;
    for (const row of tenant_rows) {
        if (row.user_id?.trim()) {
            const res = await run_reflection(row.user_id.trim(), min_override);
            total_created += res.created || 0;
        }
    }
    return { created: total_created, tenants: tenant_rows.length };
};

export const start_reflection = () => {
    if (!env.auto_reflect || timer) return;
    const int = (env.reflect_interval || 10) * 60000;
    timer = setInterval(
        () => run_reflection_all_tenants().catch((e) => console.error("[REFLECT]", e)),
        int,
    );
    console.error(`[REFLECT] Started: every ${env.reflect_interval || 10}m`);
};

export const stop_reflection = () => {
    if (timer) {
        clearInterval(timer);
        timer = null;
    }
};

File: packages/openmemory-py/src/openmemory/memory/reflect.py
import asyncio
import time
import math
import json
import logging
from typing import List, Dict, Any, Optional

from ..core.db import q, db, log_maint_op
from ..core.config import env
from ..utils.vectors import cos_sim
from .hsg import add_hsg_memory

# Ported from backend/src/memory/reflect.ts

from ..core.security import get_encryption

logger = logging.getLogger("reflect")

import re
from ..utils.vectors import cos_sim, buf_to_vec

def vec_tf(txt: str) -> List[int]:
    w = re.findall(r'\w+', txt.lower())
    uniq = sorted(list(set(w)))
    return [w.count(u) for u in uniq]

def sim_content(m1: Dict, m2: Dict, t1: str, t2: str) -> float:
    # Try semantic similarity first if vectors exist
    v1_buf = m1.get("mean_vec")
    v2_buf = m2.get("mean_vec")
    
    if v1_buf and v2_buf:
        try:
             # If it's already list (unlikely from DB raw row, but possible if processed)
             if isinstance(v1_buf, list): v1 = v1_buf
             else: v1 = buf_to_vec(v1_buf)
             
             if isinstance(v2_buf, list): v2 = v2_buf
             else: v2 = buf_to_vec(v2_buf)
             
             return cos_sim(v1, v2)
        except Exception:
             pass # Fallback to text

    # Faster text Jaccard with regex tokenization
    s1 = set(re.findall(r'\w+', t1.lower()))
    s2 = set(re.findall(r'\w+', t2.lower()))
    if not s1 or not s2: return 0.0
    
    inter = len(s1.intersection(s2))
    union = len(s1.union(s2))
    return inter / union if union > 0 else 0.0

def cluster(mems: List[Dict]) -> List[Dict]:
    cls = []
    used = set()
    
    for m in mems:
        if m["id"] in used: continue
        if m["primary_sector"] == "reflective": continue
        if m.get("meta") and "consolidated" in str(m["meta"]): continue
        
        # Helper to get plain text
        text_cache = {}
        def get_text(mem):
             if mem["id"] in text_cache: return text_cache[mem["id"]]
             try:
                 pt = get_encryption().decrypt(mem["content"])
             except Exception:
                 pt = mem["content"]
             text_cache[mem["id"]] = pt
             return pt

        text_m = get_text(m)
        c = {"mem": [m], "n": 1}
        used.add(m["id"])
        
        for o in mems:
            if o["id"] in used: continue
            if m["primary_sector"] != o["primary_sector"]: continue
            
            # Hybrid similarity: prefer vector, fallback to text
            if sim_content(m, o, text_m, get_text(o)) > 0.85: # Increased threshold for semantic
                c["mem"].append(o)
                c["n"] += 1
                used.add(o["id"])
                
        if c["n"] >= 2: cls.append(c)
        
    return cls

def calc_sal(c: Dict) -> float:
    now = time.time() * 1000
    p = c["n"] / 10.0
    
    r_sum = 0
    for m in c["mem"]:
        created = m["created_at"]
        r_sum += math.exp(-(now - created) / 43200000)
        
    r = r_sum / c["n"]
    
    # Check if any memory has 'emotional' sector in 'sectors' column (if it exists) or primary
    # TS line 66: m.sectors.includes("emotional").
    # My DB schema doesn't store 'sectors' list column directly, only primary. 
    # But `add_hsg_memory` returns `sectors`.
    # Wait, schema `memories` table DOES NOT have `sectors` column.
    # It has `primary_sector`.
    # TS `types.ts` defines `MemRow` with optional `sectors`. 
    # But `db.ts` SQLite schema doesn't have it.
    # So `m.sectors` in TS likely comes from runtime join or ignored?
    # `hsg.ts` `add_hsg_memory` returns it.
    # In `reflect.ts`, `m` comes from `q.all_mem.all`. 
    # `all_mem` select query: `select * from memories`.
    # It won't have `sectors` column.
    # So TS `m.sectors` is undefined. `includes` would throw or return false.
    # So `e` is always 0 in TS. 1:1 parity -> e=0.
    e = 0
    if c["mem"][0]["primary_sector"] == "emotional":
        e = 1.0
    else:
        for m in c["mem"]:
            meta = json.loads(m.get("meta") or "{}")
            if "emotional" in meta.get("additional_sectors", []):
                e = 0.5
                break
    
    return min(1.0, 0.6 * p + 0.3 * r + 0.1 * e)

def summ(c: Dict) -> str:
    sec = c["mem"][0]["primary_sector"]
    n = c["n"]
    # Decrypt content for summary
    enc = get_encryption()
    decoded = []
    for m in c["mem"]:
        try:
             decoded.append(enc.decrypt(m["content"]))
        except Exception:
             decoded.append(m["content"])
             
    txt = "; ".join([t[:60] for t in decoded])
    return f"{n} {sec} pattern: {txt[:200]}"

async def mark_consolidated(ids: List[str], commit: bool = True):
    if not ids: return
    now = int(time.time() * 1000)
    placeholders = ",".join(["?"] * len(ids))
    # We use a batch update to mark as consolidated
    t = q.tables
    sql = f"UPDATE {t['memories']} SET meta = json_set(COALESCE(meta, '{{}}'), '$.consolidated', true), updated_at=? WHERE id IN ({placeholders})"
    await db.async_execute(sql, (now,) + tuple(ids))
    if commit: await db.async_commit()

# boost
async def boost(ids: List[str], commit: bool = True):
    if not ids: return
    now = int(time.time() * 1000)
    placeholders = ",".join(["?"] * len(ids))
    # Batch boost salience by 10% capped at 1.0
    t = q.tables
    sql = f"UPDATE {t['memories']} SET salience = MIN(1.0, COALESCE(salience, 0) * 1.1), updated_at=? WHERE id IN ({placeholders})"
    await db.async_execute(sql, (now,) + tuple(ids))
    if commit: await db.async_commit()

async def run_reflection(user_id: Optional[str] = None) -> Dict[str, Any]:
    logger.info(f"[REFLECT] Starting reflection job for user: {user_id or 'all'}...")
    min_mems = env.reflect_min or 20
    # Increased default scan range from 100 to 500 for better pattern detection
    limit = env.reflect_limit or 500
    mems = await q.all_mem(limit, 0, user_id=user_id)
    logger.info(f"[REFLECT] Fetched {len(mems)} memories (min {min_mems})")
    
    if len(mems) < min_mems:
        logger.info("[REFLECT] Not enough memories, skipping")
        return {"created": 0, "reason": "low"}
        
    cls = cluster(mems)
    logger.info(f"[REFLECT] Clustered into {len(cls)} groups")
    
    async with db.transaction(): # Use db.transaction for async context manager
        n = 0
        for c in cls:
            txt = summ(c)
            s = calc_sal(c)
            src = [m["id"] for m in c["mem"]]
            meta = {
                "type": "auto_reflect",
                "sources": src,
                "freq": c["n"],
                "at": time.strftime("%Y-%m-%dT%H:%M:%S")
            }
            
            logger.info(f"[REFLECT] Creating reflection: {c['n']} mems, sal={s:.3f}, sec={c['mem'][0]['primary_sector']}")
            
            # Insert reflection
            # Pass commit=False to avoid nested transaction (since we are in db.transaction())
            await add_hsg_memory(txt, json.dumps(["reflect:auto"]), meta, user_id=user_id, commit=False)
            await mark_consolidated(src, commit=False)
            await boost(src, commit=False)
            n += 1
            
        if n > 0: 
            # Manually call q.ins_log or equivalent to bypass its commit?
            # Or just let log_maint_op commit at the very end.
            # log_maint_op in db.py has db.commit().
            # If I call it inside the transaction block, it will commit!
            # I should use db.execute directly for log_maint_op logic if I want atomicity.
            t = q.tables
            await db.async_execute(f"INSERT INTO {t['stats']} (type, count, ts) VALUES (?, ?, ?)", ("reflect", n, int(time.time() * 1000)))
            
    logger.info(f"[REFLECT] Job complete: created {n} reflections")
    return {"created": n, "clusters": len(cls)}

_timer_task = None

async def reflection_loop():
    interval = (env.reflect_interval or 10) * 60
    while True:
        try:
            # Multi-tenancy: Process per user
            t = q.tables
            users_rows = await db.async_fetchall(f"SELECT id FROM {t['users']}")
            user_ids = [r["id"] for r in users_rows]
            
            if not user_ids:
                # Fallback to anonymous if no users, or just skip
                await run_reflection(None)
            else:
                for uid in user_ids:
                    await run_reflection(uid)
                    await asyncio.sleep(1) # Yield
        except Exception as e:
            logger.error(f"[REFLECT] Error: {e}")
        await asyncio.sleep(interval)

def start_reflection():
    global _timer_task
    if not getattr(env, "auto_reflect", True) or _timer_task: return
    _timer_task = asyncio.create_task(reflection_loop())
    logger.info(f"[REFLECT] Started: every {env.reflect_interval or 10}m")

def stop_reflection():
    global _timer_task
    if _timer_task:
        _timer_task.cancel()
        _timer_task = None
